Company Tab and Go

nikital opened the pull request:
This pull request implements the feature discussed extensively in #526, #517, #343 and #216.

It's implemented as a separate package with minimal configuration required. To enable it just add `company-tng-frontend` to the frontend list. The code is modified from my Gist in #526 in order to enable minimum configuration for the user.

Huge thanks to @dgutov and @aaronjensen for helping me iterate on this feature 😃 

I sent a mail to assign@gnu.org, I'll let you know when the assignment bureaucracy is done.

aaronjensen commented on company-tng.el:
```diff
@@ -0,0 +1,128 @@
+;;
+;; Usage:
+;;
+;; Add `company-tng-frontent' to `company-frontends':
```
frontend

dgutov commented on company-tng.el:
```diff
@@ -0,0 +1,128 @@
+;;; company-tng.el --- company-mode frontend that inserts selected candidates
```
We'll also need a NEWS entry.

nikital commented:
I fixed your comments and force-pushed them to nikital:company-tng
Cheers 🥂

dgutov commented on NEWS.md:
```diff
@@ -3,6 +3,7 @@
 ## Next
 
 * New user option `company-tooltip-maximum-width`.
+* New frontend `company-tng'.
```
It's more than just a frontend, right? Let's call it "new subpackage", or a "new feature" (with an explanation).

I think NEWS is one of the main sources of users finding out about new stuff in company-mode.

dgutov commented:
Thanks, but this comment is still not addresses, I think:

> Could we maybe have some other summary? "frontend that inserts" doesn't make a lot of sense to me.

aaronjensen commented:
It might be nice to add an auto loaded command that would install company-tng with the recommended bindings.

nikital commented:
@aaronjensen Nice idea, I added (company-tng-configure-default) to help new users get started

@dgutov I edited the NEWS file and added more details. Regarding the comment at the top of the file: Last time I changed it to "company-mode frontend that inserts a candidate into the buffer as soon as it's selected, Vim style", but I have no further ideas to improve it. If you have anything in mind please share and I'll put it in.
Additionally, FYI I finished the copyright assignment for GNU Emacs.

dgutov commented:
> If you have anything in mind please share and I'll put it in.

Maybe something like "collection of settings for tab-only interaction"? It's a bit vague, though.

> Additionally, FYI I finished the copyright assignment for GNU Emacs.

Great! I'll merge this soon, then.

nikital commented:
> Maybe something like "collection of settings for tab-only interaction"? It's a bit vague, though.

I like it less, I think that describing a package by it's default keybinding is not very informative and I prefer to describe it by it's functionality (inserts a candidate into the buffer as soon as it's selected).

In any case, I'm really not into bikeshedding over this 😄, so if you prefer "collection of settings for tab-only interaction" I don't mind at all, just let me know and I'll update the pull request to:

    ;;; company-tng.el --- collection of settings for tab-only interaction

Have a good weekend!

dgutov commented:
@nikital 

> just let me know and I'll update the pull request to

Eh, it's okay. Still undecided, but I can update the summary later anyway.

Could you give me your copyright assignment number? And/or the email corresponding to it. I don't see it in the internal list, but maybe it just needs updating.

nikital commented:
Sorry for the delay, the assignment is finally complete. (I misunderstood the FSF process so I thought it was done before they mailed me the documents back)

nikital commented:
Hi, I pushed a bug fix. The completion wasn't displayed in the buffer when `company-prefix` was empty...

dgutov commented:
Thanks! Merging; I'll go over the documentation changes now.

nikital commented:
@dgutov Great to see it merged! It's my first non-trivial contribution to an open source project, so thanks for helping me take this step :)

dgutov commented:
@nikital You're welcome, and thanks again ;)

jojojames commented:
Thanks @nikital and @dgutov.

Leaving some snips here from trying it out in case it's useful for someone else.

Noticed there's an extra preview added when combining tng with preview frontend.
```
(setq company-frontends
      '(company-tng-frontend
        company-pseudo-tooltip-unless-just-one-frontend
        company-echo-metadata-frontend
        company-preview-frontend
        company-quickhelp-frontend))

;; Redefining like this works for me.
(defun company-preview-frontend (command)
  "`company-mode' frontend showing the selection as if it had been inserted."
  (pcase command
    (`pre-command (company-preview-hide))
    (`post-command
     (unless (and company-selection-changed
                  (memq 'company-tng-frontend company-frontends))
       (company-preview-show-at-point (point)
                                      (nth company-selection company-candidates))))
    (`hide (company-preview-hide))))
```

Also became tricky to expand snippets when there's more than one. Using RET for that seems like a good idea.

```
(defun +company-expand-yasnippet-or-return ()
    "Expand yas template or call RET normally."
    (interactive)
    (if (and company-active-map
             company-selection-changed
             (car company-candidates)
             (get-text-property 0 'yas-template (car company-candidates)))
        (call-interactively #'company-complete-selection)
      (when company-selection-changed
        (company-complete-selection))
      (let ((company-active-map))
        (call-interactively (key-binding (kbd "RET"))))))

  (define-key company-active-map [return] '+company-expand-yasnippet-or-return)
  (define-key company-active-map (kbd "RET") '+company-expand-yasnippet-or-return)
```

dgutov commented:
@jojojames

> Noticed there's an extra preview added when combining tng with preview frontend

What did you do for this to happen?

But that reminds me... @nikital, try this scenario: in an `emacs-lisp-mode` buffer, type `defal` (it should get you a single completion), then press TAB. What do you expect to happen, and what do you see?

dgutov commented:
@jojojames 

> Also became tricky to expand snippets when there's more than one

Less obvious maybe, but tricky how? You select the desired snippet using TAB the same way as you pick a non-snippet completion.

jojojames commented:
>> What did you do for this to happen?
I type a few characters (with those frontends set), a popup appears with list of completions. I press tab once, tng draws the completion into view. Preview frontend draws the preview after the current completion. Maybe it's my set up also.

>> Less obvious maybe, but tricky how? You select the desired snippet using TAB the same way as you pick a non-snippet completion.
Hmnn, this is the scenario I was thinking of.

Popup
[Completion 1
Completion 2
Completion 3 (yasnippet)]

I can select the snippet by tabbing to completion 3 but not sure how to expand it? If I press tab again, it'll loop the selection. If I press enter, it seems to expand and insert a new line into the snippet.

aaronjensen commented:
> I can select the snippet by tabbing to completion 3 but not sure how to expand it?

You'll need a specific binding to expand it, maybe `M-/`. The concept of tng and snippet expansion are a bit incompatible, but adding an additional binding just for the snippet expansion should work.

jojojames commented:
> You'll need a specific binding to expand it, maybe M-/. The concept of tng and snippet expansion are a bit incompatible, but adding an additional binding just for the snippet expansion should work.
Yup! Using the RET key and setting up a fallback keybind seems to be DWIM instead of using something like M-/.

dgutov commented:
> I can select the snippet by tabbing to completion 3 but not sure how to expand it?

Why not start typing whatever you would have been typing after the snippet is expanded? If it's expands into a method call, start typing the first argument. I agree it's not intuitive (and I've love to hear the ideas from the Vim/ycmd world), but it should work and be efficient enough, as soon as you get used to it.

Ultimately, though, `company-tng-frontend` might even automatically "expand" the snippets, just like it "inserts" completions now.

drot commented:
> Preview frontend draws the preview after the current completion.

Can confirm. Starting with emacs -Q and evaluating the following in order:


![company](https://user-images.githubusercontent.com/85483/30821887-2e34ac84-a227-11e7-8926-20721caecae8.png)

Last line is what happens when you press TAB while the preview frontend is active.

dgutov commented:
@nikital In short, I think when there is only one candidate in the list, `company-tng--allow-unselected` should call `company-complete-selection` somehow (not sure how to implement this best with the current advice approach). That would solve the problem with the frontends automatically.

nikital commented:
### TNG + Preview
I see the problem. I think that we should change the default configuration for TNG:

    (defun company-tng-configure-default ()
      "Applies the default configuration to enable company-tng."
      (setq company-frontends '(company-tng-frontend
                                company-pseudo-tooltip-frontend
                                company-echo-metadata-frontend))
      ...

1. It doesn't include the preview frontend. I think that preview is not needed when TNG is active because TNG already does the same thing (previews selected candidates in-buffer)
2. It uses `company-pseudo-tooltip-frontend` instead of `company-pseudo-tooltip-unless-just-one-frontend`. Vim's/ycm's default behavior is to show the tooltip even if there is one candidate and the user can just hit TAB to select it. I think that it makes sense because one candidate and many candidates are presented in the same way and TAB work consistently between them.

### Snippets
I'm in @aaronjensen's camp on snippet expansion and I prefer to use a dedicated hotkey instead of multiplexing it on another key, but I'm glad that @jojojames's `+company-expand-yasnippet-or-return` works as well!

> Ultimately, though, company-tng-frontend might even automatically "expand" the snippets, just like it "inserts" completions now.

That's a nice concept, I like the general idea but I'm not familiar with yasnippet at all so I have no feedback on possible implementation... I also wonder what should happen during regular completion if the user TABs over a really big snippet. Will we just temporarily preview the entire expanded snippet in the buffer? Could create an unpleasant visual flash...

jojojames commented:
> Why not start typing whatever you would have been typing after the snippet is expanded?

Hmnn, not sure I understand this workflow. I do like the RET to confirm the snippet expansion though. So I'll just stick with that.

> Ultimately, though, company-tng-frontend might even automatically "expand" the snippets, just like it "inserts" completions now.

This seems cool but I think yasnippet also takes control of the tab key when it's active so it might be a little awkward cycling through completions as well as going to the next part of the snippet.

> It doesn't include the preview frontend. I think that preview is not needed when TNG is active because TNG already does the same thing (previews selected candidates in-buffer)

I like the preview while typing and I think TNG only comes into play after the first selection. It'd be nice if the frontends played nice with each other when combining them.

>  It uses company-pseudo-tooltip-frontend instead of company-pseudo-tooltip-unless-just-one-frontend. Vim's/ycm's default behavior is to show the tooltip even if there is one candidate and the user can just hit TAB to select it. I think that it makes sense because one candidate and many candidates are presented in the same way and TAB work consistently between them.

Changing the defaults is fine but I actually think it's a limitation of vim (IIRC it can't show a preview like Emacs does). This might be a case of reproducing ycmd/vim behavior faithfully or doing what's best (whatever that means).

>  I agree it's not intuitive (and I've love to hear the ideas from the Vim/ycmd world)

Having used vim for a while with ycmd, I've never found ycmd+ultisnips (not sure about new packages) to have great synergy.

dgutov commented:
@nikital 

> we should change the default configuration for TNG

If you're confident in that solution, please go ahead and make a PR.

> what should happen during regular completion if the user TABs over a really big snippet

We could truncate it to the first line, maybe (with an ellipsis as a hint), as one option.

> I'm not familiar with yasnippet at all so I have no feedback on possible implementation

Neither am I familiar with it to the extent required.

dgutov commented:
@jojojames

> Hmnn, not sure I understand this workflow.

Select the completion that comes from `company-yasnippet` and press some key like SPC. But I agree that it's awkward, not seeing where that space will end up.

> it might be a little awkward cycling through completions as well as going to the next part of the snippet

Good point. Ideas welcome here as well.

But using some extra key (RET or otherwise) seems inevitable indeed.

nikital commented:
@jojojames 
> Changing the defaults is fine but I actually think it's a limitation of vim (IIRC it can't show a preview like Emacs does). This might be a case of reproducing ycmd/vim behavior faithfully or doing what's best (whatever that means).

I'm don't think we need to reproduce Vim, but for now sticking to Vim default is the easy fix that I PRed in #712.

If you want to integrate preview into the mix, I would start with binding tab to something like this and see if it behaves intuitively (didn't test it, just wrote it off the top of my head):
```
(defun company-select-next-or-complete (&optional arg)
  "Select the next candidate if more than one, else complete.

With ARG, move by that many elements."
  (interactive "p")
  (if (> company-candidates-length 1)
      (company-select-next arg)
    (company-complete-selection)))

```

rswgnu commented:
Although C-g can be used to abort company completion, often one's muscle memory (from finishing prompted for arguments or entry of forms) will reach for RETURN to finish things up.  Since the default configuration of tng does not bind RETURN to anything, why not make it run company-abort.  Then if you don't want to select a completion from the popup, a quick RETURN
makes it disappear.  Also this requires only a one finger press instead of two.

jojojames commented:
I believe that's one of the biggest benefits of TNG, that RET isn't bound to anything.

You can bind conpany-abort to company active map if you want though.

rswgnu commented:
Actually, I don't want RETURN to do a company-abort (because that removes any selected completion).  I just want it to leave whatever completion has been added to the buffer and remove the company completions popup and not do anything else.  Is there an existing command to use for that?

aaronjensen commented:
`keyboard-quit`?

rswgnu commented:
No, that does not cleanly terminate from the company popup overlay.  We need a command that leaves any previously selected completion in place, does not select one unless the user has manually selected one and then cleanly terminates from company completion mode.

rswgnu commented:
Really the company-finish function should be a command which inserts into the buffer any partial or full completion that has been selected (I think tng just adds an overlay with the completion and doesn't do an insertion) but should not select the first completion entry if the user has not made such a selection.

rswgnu commented:
This is what I had to write to get the above behavior.  Feel free to add it and make any modifications you may require.  I have not yet found any problems with it, though.

```
(defun company-quit ()
  "Insert any selected completion and quit completing."
  (interactive)
  (when (and company-selection-changed company--manual-action
	     (boundp 'company-tng--overlay) company-tng--overlay)
    (company--insert-candidate
     (nth company-selection company-candidates)))
  (company-cancel))
```

rswgnu commented:
This would be simpler if company-selection were not set to the 0th item in the candidates list prior to anything being selected.  Maybe it could be nil at that point.  Then it could be tested as to whether anything is selected or not.  Or is there a simpler test for that?  And how does one de-select any item but leave the selection popup on screen?

rswgnu commented:
Wait for an update on this code; I did find an issue.

rswgnu commented:
The code does seem to work fine here.

nikital commented:
I think it's a good idea to C-g's behavior consistent with ESC in Vim/Evil (leave the inserted text). I didn't test it extensively, but I think @rswgnu's code can be simplified to:
```elisp
(defun company-tng-complete ()
  (interactive)
  (if company-selection-changed
      (company-complete-selection)
    (company-abort)))

(define-key company-active-map (kbd "C-g") 'company-tng-complete)
```
I will test it for some time on my own config to catch possible problems and then open pull request. Comments are welcome :)

jojojames commented:
@nikital I've always thought of C-g as 'get out of jail' and not 'execute some custom code related to the current package/keymap that's related to aborting'.

I'm not sure it's so important to match C-g with vim/evil's in the first place since vim/evil don't use C-g as an escape key anyways.

nikital commented:
I understand what you mean.

However, because in company-tng the candidate gets inserted into the buffer as soon as it's selected, it seems to me that leaving the candidate inside the buffer is the better "get out of jail" experience than deleting a previously inserted candidate.

jojojames commented:
I've ben using C-g to be more of a 'cancel all completions *even* if I've selected one'.

Sometimes I scroll through the list just to see what's in it (not uncommon) but aren't interested in completing anything.

dgutov commented:
> leaving the candidate inside the buffer is the better "get out of jail" experience

I disagree.

You can bind that to RET, though, if that's the consensus.

nikital commented:
I see... I didn't think about your use case (just scrolling through
candidates and then aborting if you don't find any), it makes sense to
leave the current default behavior then.

Regarding RET, I agree with Aaron that having it unbound is very important,
especially in languages such as lisp or python that have no semicolons at
the end of the line, which makes it likely to have the tooltip active when
you are at the end of the line and you want RET to actually insert a new
line.

I think the consensus is to leave things as is for now :)

On Dec 5, 2017 02:38, "Dmitry Gutov" <notifications@github.com> wrote:

> leaving the candidate inside the buffer is the better "get out of jail"
> experience
>
> I disagree.
>
> You can bind that to RET, though, if that's the consensus.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/company-mode/company-mode/pull/706#issuecomment-349264793>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AATmqtouQXly_7kRNrhFfk-mekpep72iks5s9R0ngaJpZM4PLCbu>
> .
>

rswgnu commented:
On Tue, Dec 5, 2017 at 10:26 AM, Nikita Leshenko <notifications@github.com>
wrote:

> I see... I didn't think about your use case (just scrolling through
> candidates and then aborting if you don't find any), it makes sense to
> leave the current default behavior then.
>

​Yes, C-g works well as a completion abort.​

> ​​
>
> ​​
> Regarding RET, I agree with Aaron that having it unbound is very important,
> ​​
> especially in languages such as lisp or python that have no semicolons at
> ​​
> the end of the line, which makes it likely to have the tooltip active when
> ​​
> you are at the end of the line and you want RET to actually insert a new
> ​​
> line.
>

​That is certainly one possibility.  But generally when a selectable
popup is on-screen​, one must deal with it before going on to do
other editing.  Since RET selects an item when over one and removes
the popup, it makes sense that it would remove the popup without selecting
one when not over and item and then like all keys bound in
company-active-mode
return to its normal binding afterwards.  I use Python and Lisp all the time
and have no problem having RET set to this behavior.

Similarly, C-n and C-p are much more natural for touch typists to use to
move through
the items and more consistent with typical Emacs bindings than using M-n
and M-p (I am
suggesting these bindings be made).  You don't move around in a buffer in
the middle of
completing because then completion won't work properly.  With this setup,
C-n, C-p, C-g
and RET do everything you need without use of any more hard-to-reach keys
and that would
improve speed more than what their current uses do during company mode.

​
>
> ​​
> I think the co
> ​​
> nsensus is to leave t
> ​​
> hings as is for now :)


​If so, so be it, I just want to stimulate discussion and thought about
possible improvements.

​

aaronjensen commented:
> it makes sense that it would remove the popup without selecting
> one when not over and item and then like all keys bound in
> company-active-mode

@rswgnu I believe you're describing (more or less) the original company behavior. company-tng was requested and built by people who want exactly what you are saying you don't want. We want to be able to have the selection of the item (via tab/shift-tab) be the confirmation. The extra keypress to select is specifically what we don't want. 

It sounds like you maybe want some hybrid, but I'm not sure exactly what it is that you want that the original company behavior does not support.

rswgnu commented:
On Tue, Dec 5, 2017 at 11:35 AM, Aaron Jensen <notifications@github.com>
wrote:

> it makes sense that it would remove the popup without selecting
> one when not over and item and then like all keys bound in
> company-active-mode
>
> @rswgnu <https://github.com/rswgnu> I believe you're describing (more or
> less) the original company behavior. company-tng was requested and built by
> people who want exactly what you are saying you don't want. We want to be
> able to have the selection of the item (via tab/shift-tab) be the
> confirmation. The extra keypress to select is specifically what we don't
> want.
>
> It sounds like you maybe want some hybrid, but I'm not sure exactly what
> it is that you want that the original company behavior does not support.
>
​The issues I am discussing are:
  What does RET do when no popup completion is visibly selected?  I want it
to quit from company-mode and leave any partial completion that was done
already in the buffer; nothing more or less.
  If RET is pressed on a popup completion, it should be inserted into the
buffer as the completion and company-mode quit.  I believe the default
behavior of company-mode is to invisibly select
  the first potential completion item, so an immediate RET adds that item
as the completion in the buffer.  I never want that behavior and I find it
wastes a lot of time correcting things.
  Often I may want to look at completions without doing any completion.  I
find RET much simpler to use than C-g in such cases.  I want to use C-g
only to abort after seeing some completion
  added to the buffer.

  I also suggested C-p and C-n be bound to move through popup completion
items.

  I like having TAB perform completions as well, so I use company-tng.  But
TAB just moves between popup items and displays each potential completion
in the buffer; it doesn't
  terminate the popup or company-mode.  How could it do that and still move
through the popup selections?  You still have to press RET or some
non-company-mode key to have the
  selection permanently added to the buffer, right?

Bob

aaronjensen commented:
> What does RET do when no popup completion is visibly selected?

Under company-tng, it closes the popup and inserts a return. This is the behavior that other editors/completion engines have that company-tng is emulating. 

If you want different behavior you'll probably need to write it. It doesn't make sense to change company-tng on this and it'd be up to @dgutov if he wants to change the default behavior of RET, which, at the moment, does not "invisibly select the first potential candidate" as that candidate is already selected. Company selects it immediately when it pops up. company-tng doesn't do this, which is perhaps why you're thinking that it invisibly selects it.



> I also suggested C-p and C-n be bound to move through popup completion
> items.

You can bind these in your init. I have them bound to this as well, I don't know where that's coming from--perhaps spacemacs, but I use tab/shift-tab generally any way. I may actually unbind them because I want to free those up to move around.




> I like having TAB perform completions as well, so I use company-tng.  But
> TAB just moves between popup items and displays each potential completion
> in the buffer; it doesn't
>   terminate the popup or company-mode.  How could it do that and still move
> through the popup selections?  You still have to press RET or some
> non-company-mode key to have the
>   selection permanently added to the buffer, right?

That's right, and this is exactly the behavior that all of who asked for and contributed to company-tng wanted.

Again, it sounds like you want a hybrid, specifically you want to be able to press RET when nothing has been selected to dismiss company-mode or to (actually) insert the candidate if one is selected without also inserting a newline. It looks like you wrote a function to do what you want. Perhaps if you want to pull request that function @dgutov may be interested in having it built-in so others could bind it to whatever they wanted, if they wanted to be used in conjunction with `company-tng`. I would say it should not be the default, however.

rswgnu commented:
On Tue, Dec 5, 2017 at 2:15 PM, Aaron Jensen <notifications@github.com>
wrote:

> Company selects
> ​[the first completion candidate]​
> immediately when it pops up. company-tng doesn't do this, which is perhaps
> why you're thinking that it invisibly selects it.
>
> ​With company-tng loaded, before I customized anything, once a completions
list popped up, nothing was selected, but if I pressed return, the first
candidate was inserted into the buffer.
So that did appear to be as an invisible selection of the first candidate,
as you noted.

> Aga
> ​​
> in, it sounds like you want a hybrid, specifically you want to be able to
> press RET when nothing has been selected to dismiss company-mode or to
> (actually) insert the candidate if one is selected without also inserting a
> newline.
>
> ​Yes, said more simply, I want consistent behavior which is that RET
inserts any selected candidate and quits from company-mode.  company-tng is
not consistent since it inserts a non-selected candidate when RET is
pressed prior to selecting any candidate.
​​
It lo
​​
ok
​​
s like you wrote a function to do what you want.

Yes, I am only writing to encourage a change in defaults.

> Perhaps if you want to pull request that function @dgutov
> <https://github.com/dgutov> may be intereste
> ​​
> d in having it built-in so others could bind it to whatever they wanted,
> if they wanted to be used in conjunction with company-tng. I would say it
> should not be the default, however.
>
> ​That's a reasonable idea but based on past experience I would ask Dimitri
to express some interest in this before I do anything further with it.  If
he is interested, I will do it.

Thanks for your thoughts on this.
​​

aaronjensen commented:
> With company-tng loaded, before I customized anything, once a completions
> list popped up, nothing was selected, but if I pressed return, the first
> candidate was inserted into the buffer.
> So that did appear to be as an invisible selection of the first candidate,
> as you noted.

did you invoke `(company-tng-configure-default)`? It unmaps RET and it should not do what you're describing.

rswgnu commented:
On Tue, Dec 5, 2017 at 4:57 PM, Aaron Jensen <notifications@github.com>
wrote:

> With company-tng loaded, before I customized anything, once a completions
> list popped up, nothing was selected, but if I pressed return, the first
> candidate was inserted into the buffer.
> So that did appear to be as an invisible selection of the first candidate,
> as you noted.
>
> did you invoke (company-tng-configure-default)?
>
​Yes.​

> It
> ​​
> u
> ​​
> nmaps RET and it should not do what you're describing.
>
> ​I saw this behavior many times before I fixed it.  I should note that
sometimes I am using the jedi backend for Python and its use of deferred
function calls leads to all sorts of oddness at times.

But I basically loaded company and company-tng out-of-the-box, set them up
as recommended, turned on global company mode, and bound the following keys
like so:

  (global-set-key [remap completion-at-point] 'company-complete)
  (global-set-key [remap complete-symbol] 'company-complete)
  (global-set-key [remap indent-for-tab-command]
'company-indent-or-complete-common)


With that I observed the problematic RET behavior.  To fix it, in
company-active-map, I had to bind RET to this function that I wrote:

  (defun company-quit ()
    "Insert any selected completion and quit completing."
    (interactive)
    (when (and company-selection-changed company--manual-action
       (boundp 'company-tng--overlay) company-tng--overlay)
      (company--insert-candidate
       (nth company-selection company-candidates)))
    (company-cancel))

Bob

jojojames commented:
company-active-map

company-backends

company-frontends

company-diag (this is a function)

It might be more productive to give values of those^ variables.

That global-set-key for the indent-for-tab-command looks suspicious though. Why do you need to rebind those keys? Part of the tng-setup is rebinding keys for you.

rswgnu commented:
On Tue, Dec 5, 2017 at 6:58 PM, James N <notifications@github.com> wrote:

> That global-set-key for the indent-for-tab-command looks suspicious
> though. Why do you need to rebind those keys? Part of the tng-setup is
> rebinding keys for you.
>
​I just tested without indent-for-tab-command rebound and in that case, I
don't get a completions popup when I press tab after:
​

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/company-mode/company-mode/pull/706#issuecomment-349482430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATtnj6venCD_yoG5QNB_WhpyD2s47tyTks5s9di4gaJpZM4PLCbu>
> .
>

jojojames commented:
I'm coming to the conclusion that the behavior you're seeing might not be how company-tng is expected to behave?

It might be better to debug with emacs -q and/or provide those ^values from above and/or bisect your config and see what's going on.

>  I think the consensus is to leave things as is for now :)

+1

rswgnu commented:
On Tue, Dec 5, 2017 at 8:09 PM, James N <notifications@github.com> wrote:

> I'm coming to the conclusion that the behavior you're seeing might not be
> how company-tng is expected to behave?
>
​So by default with tng enabled, if no completion is yet selected, pressing
RET is supposed to simply quit from company-mode and add a newline to the
buffer, right?​

​​

That is not the behavior I remember seeing though.

> It migh
> ​​
> t be better to debug with emacs -q and/or provide those ^values from above
> and/or bisect your config and
> ​​
> see what's going on.
> ​​
>
> ​Yes, I had started doing that but haven't had time to finish tracing
through things.

rswgnu commented:
On Tue, Dec 5, 2017 at 6:58 PM, James N <notifications@github.com> wrote:

> That global-set-key for the indent-for-tab-command looks suspicious
> though. Why do you need to rebind those keys? Part of the tng-setup is
> rebinding keys for you.
>

I j
​​
ust
​​
tested without indent-for-tab-command rebound and in that case, I don't get
a completions popup when I press tab
​in python-mode ​
after:
​
sys.​
​With that key remapped, I see the expected popup showing features of the
sys module.​
​
Bob

nikital commented:
@rswgnu I couldn't understand your last message because your email client messed up the formatting. Could you fix it or use Github's web interface?

> So by default with tng enabled, if no completion is yet selected, pressing RET is supposed to simply quit from company-mode and add a newline to the buffer, right?​

Yes. Since the idea of company-tng is to make TAB/S-TAB cycle through candidates and insert them as they are selected, you can, theoretically, unbind all company-mode keys except TAB and it would still work. RET isn't a special key in company-tng, that's why it's unbound by default.

As other have mentioned, apparently you have another package that binds RET back and creates this weird behavior. The "unselected" state is something specific to company-tng, other packages will get confused by it and might insert the first candidate even if it's not selected. You should unbind RET if you want a good company-tng experience.

Also, if you can open a new issue with a minimal config that reproduces your out-of-the-box experience I might be able to modify company-tng to provide a better integration (For example maybe we should advice `company-complete-selection` to make it aware of the unselected state when company-tng is active). Please @ mention me on the new issue.

terlar commented:
First of all really insightful discussions, and it had lead me into refining my setup and really understand what I would want.

I almost have what I want with `company-tng` and `company-preview`, I want to use `company-preview` before anything is selected by `company-pseudo-tooltip`. I think it is most easily explainable by a short gif.

My current configuration:
```
(defun company-preview-if-none-selected-frontend (command)
  "`company-preview-frontend', but only when nothing is selected."
  (message "%s" company-selection)
  (unless (and (eq command 'post-command)
               (> company-selection 0))
    (company-preview-frontend command)))
(setq company-frontends
      '(company-tng-frontend
        company-pseudo-tooltip-unless-just-one-frontend
        company-preview-if-none-selected-frontend
        company-echo-metadata-frontend)
```

![company-demo](https://user-images.githubusercontent.com/280235/33798764-d6f0df1e-dd1e-11e7-981d-2e68915f04ac.gif)

The reason why I want to have the preview is that I find it easier to see, ah this is the text I want and then I just accept directly, however when I start selecting from the menu I am fine with that insertion. So I wonder how to solve a preview that only is shown before you enter the pseudo-tooltip box. As shown by the video there are two different states where both have `company-selection` = `0`. You can see it as I select-previous out of the box and nothing is selected, then techincally it is still `0` and that is the case when I would want `company-preview` to be active.

Is it possible to differentiate by those states, or is it only internal to the pseudo tooltip?

jojojames commented:
@terlar What theme/font are you using?

Also, did you try my workaround (mentioned above.) That should fix your problem.

I do think we should try to get company-preview working with company-tng. I've seen this asked a couple times now and it's definitely a source of confusion.

terlar commented:
@jojojames Thank you, that works very well. I just ignored it since there was a lot of discussion of the problem after your function post with other solutions, so I assumed it didn't work.

I rewrote it as such (to not having to reproduce logic):
```elisp
  (defun company-preview-if-not-tng-frontend (command)
    "`company-preview-frontend', but not when tng is active."
    (unless (and (eq command 'post-command)
                 company-selection-changed
                 (memq 'company-tng-frontend company-frontends))
      (company-preview-frontend command)))
```

@jojojames The theme I am using is [tao-theme](https://github.com/11111000000/tao-theme-emacs) using the `tao-yang` version with some [custom overrides](https://github.com/terlar/emacs.d/blob/dcf455288ea834beb71a43db3d29c540c91af97e/lisp/local-theme.el). The font is [Fira Mono](https://github.com/mozilla/Fira). :smile_cat:
